sklearn/_fast_gradient_boosting/gradient_boosting.py

```diff
@@ -0,0 +1,757 @@
+            # stratify for classification
+            stratify = y if hasattr(self.loss_, 'predict_proba') else None
+
+            X_binned_train, X_binned_val, y_train, y_val = train_test_split(
+                X_binned, y, test_size=self.validation_fraction,
+                stratify=stratify, random_state=rng)
+            if X_binned_train.size == 0 or X_binned_val.size == 0:
+                raise ValueError(
+                    'Not enough data (n_samples={}) to '
+                    'perform early stopping with validation_fraction='
+                    '{}. Use more training data or '
+                    'adjust validation_fraction.'.format(
+                        X_binned.shape[0],
+                        self.validation_fraction)
+                )
+            # Predicting is faster of C-contiguous arrays, training is faster
+            # on Fortran arrays.
+            X_binned_val = np.ascontiguousarray(X_binned_val)
+            X_binned_train = np.asfortranarray(X_binned_train)
+        else:
+            X_binned_train, y_train = X_binned, y
+            X_binned_val, y_val = None, None
+
+        # Subsample the training set for early stopping and score monitoring
+        if self.do_early_stopping_:
+            subsample_size = 10000  # should we expose this parameter?
+            indices = np.arange(X_binned_train.shape[0])
+            if X_binned_train.shape[0] > subsample_size:
+                indices = rng.choice(indices, subsample_size)
+            X_binned_small_train = X_binned_train[indices]
+            y_small_train = y_train[indices]
+            # Predicting is faster on C-contiguous arrays.
+            X_binned_small_train = np.ascontiguousarray(X_binned_small_train)
+
+        if self.verbose:
+            print("Fitting gradient boosted rounds:")
+
+        # initialize raw_predictions: those are the accumulated values
+        # predicted by the trees for the training data. raw_predictions has
+        # shape (n_trees_per_iteration, n_samples) where
+        # n_trees_per_iterations is n_classes in multiclass classification,
+        # else 1.
+        n_samples = X_binned_train.shape[0]
+        self._baseline_prediction = self.loss_.get_baseline_prediction(
+            y_train, self._n_trees_per_iteration)
+        raw_predictions = np.zeros(
+            shape=(self._n_trees_per_iteration, n_samples),
+            dtype=self._baseline_prediction.dtype
+        )
+        raw_predictions += self._baseline_prediction
+
+        # initialize gradients and hessians (empty arrays).
+        # shape = (n_trees_per_iteration, n_samples).
+        gradients, hessians = self.loss_.init_gradients_and_hessians(
+            n_samples=n_samples,
+            prediction_dim=self._n_trees_per_iteration
+        )
+
+        # estimators_ is a matrix (list of lists) of TreePredictor objects
+        # with shape (n_iter_, n_trees_per_iteration)
+        self.estimators_ = estimators = []
```

Comment:
I agree but OTOH don't we want to have an API as close as possible to the current GBDTs? That's why I changed the names.

Reply:
No because it would not make sense and induce confusion. Calling a predictor an estimator is too misleading.

Unless we want to make it possible to have single tree histogram based estimators but I thought it would be out of the scope of this PR.

Reply:
>it would be out of the scope of this PR.

Yes definitely

@jnothman @amueller are you OK to use different parameters / attribute names from the current GBDT estimators as suggested here?